sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala

```diff
@@ -22,10 +22,40 @@ import java.time.chrono.IsoChronology
 import java.time.format.{DateTimeFormatter, DateTimeFormatterBuilder, ResolverStyle}
 import java.time.temporal.{ChronoField, TemporalAccessor, TemporalQueries}
 import java.util.Locale
+import java.util.concurrent.ConcurrentHashMap
 
 trait DateTimeFormatterHelper {
+  protected def toInstantWithZoneId(temporalAccessor: TemporalAccessor, zoneId: ZoneId): Instant = {
+    val localTime = if (temporalAccessor.query(TemporalQueries.localTime) == null) {
+      LocalTime.ofNanoOfDay(0)
+    } else {
+      LocalTime.from(temporalAccessor)
+    }
+    val localDate = LocalDate.from(temporalAccessor)
+    val localDateTime = LocalDateTime.of(localDate, localTime)
+    val zonedDateTime = ZonedDateTime.of(localDateTime, zoneId)
+    Instant.from(zonedDateTime)
+  }
+}
+
+object DateTimeFormatterHelper {
+  private val cache = new ConcurrentHashMap[(String, Locale), DateTimeFormatter]()
```

Comment:
Do we need to consider cleaning up old entries in this map?

Reply:
Yea, I was wondering about that too. This combinations looks huge.

Reply:
Also answering my own question: the formatter is thread-safe, so this is fine.
I agree that this cache won't grow large as I can't imagine an app using more than a handful of distinct patterns and locales.

Reply:
Yea but this can be performed row-based operation as a corner case. For instance, an expression that allows different formats for each value can make this cache grows a lot. It's unlikely but pretty possible. Also, most importantly currently it doesn't look super useful for the same reason of https://github.com/apache/spark/pull/23462#issuecomment-451714252

Reply:
True. This could wrap some weak-ref based cache if we needed to.
Good point about whether this is really created a lot.

Reply:
shall we use `com.google.common.cache.Cache`? It's used in several places inside Spark.

Reply:
If we have significant concern of infinitely cache growing, I will switch on fixed size LRU cache here.

Reply:
I think it makes sense to restrict growing of the cache. I replaced it by Guava's cache.